<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/xhtml;charset=UTF-8"/>
<meta http-equiv="X-UA-Compatible" content="IE=9"/>
<meta name="generator" content="Doxygen 1.8.6"/>
<title>BSG: HaplotypeScorer Class Reference</title>
<link href="tabs.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="jquery.js"></script>
<script type="text/javascript" src="dynsections.js"></script>
<link href="navtree.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="resize.js"></script>
<script type="text/javascript" src="navtree.js"></script>
<script type="text/javascript">
  $(document).ready(initResizable);
  $(window).load(resizeHeight);
</script>
<link href="search/search.css" rel="stylesheet" type="text/css"/>
<script type="text/javascript" src="search/search.js"></script>
<script type="text/javascript">
  $(document).ready(function() { searchBox.OnSelectItem(0); });
</script>
<link href="doxygen.css" rel="stylesheet" type="text/css" />
</head>
<body>
<div id="top"><!-- do not remove this div, it is closed by doxygen! -->
<div id="titlearea">
<table cellspacing="0" cellpadding="0">
 <tbody>
 <tr style="height: 56px;">
  <td style="padding-left: 0.5em;">
   <div id="projectname">BSG
   </div>
  </td>
 </tr>
 </tbody>
</table>
</div>
<!-- end header part -->
<!-- Generated by Doxygen 1.8.6 -->
<script type="text/javascript">
var searchBox = new SearchBox("searchBox", "search",false,'Search');
</script>
  <div id="navrow1" class="tabs">
    <ul class="tablist">
      <li><a href="index.html"><span>Main&#160;Page</span></a></li>
      <li><a href="pages.html"><span>Related&#160;Pages</span></a></li>
      <li class="current"><a href="annotated.html"><span>Classes</span></a></li>
      <li><a href="files.html"><span>Files</span></a></li>
      <li>
        <div id="MSearchBox" class="MSearchBoxInactive">
        <span class="left">
          <img id="MSearchSelect" src="search/mag_sel.png"
               onmouseover="return searchBox.OnSearchSelectShow()"
               onmouseout="return searchBox.OnSearchSelectHide()"
               alt=""/>
          <input type="text" id="MSearchField" value="Search" accesskey="S"
               onfocus="searchBox.OnSearchFieldFocus(true)" 
               onblur="searchBox.OnSearchFieldFocus(false)" 
               onkeyup="searchBox.OnSearchFieldChange(event)"/>
          </span><span class="right">
            <a id="MSearchClose" href="javascript:searchBox.CloseResultsWindow()"><img id="MSearchCloseImg" border="0" src="search/close.png" alt=""/></a>
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div id="navrow2" class="tabs2">
    <ul class="tablist">
      <li><a href="annotated.html"><span>Class&#160;List</span></a></li>
      <li><a href="classes.html"><span>Class&#160;Index</span></a></li>
      <li><a href="inherits.html"><span>Class&#160;Hierarchy</span></a></li>
      <li><a href="functions.html"><span>Class&#160;Members</span></a></li>
    </ul>
  </div>
</div><!-- top -->
<div id="side-nav" class="ui-resizable side-nav-resizable">
  <div id="nav-tree">
    <div id="nav-tree-contents">
      <div id="nav-sync" class="sync"></div>
    </div>
  </div>
  <div id="splitbar" style="-moz-user-select:none;" 
       class="ui-resizable-handle">
  </div>
</div>
<script type="text/javascript">
$(document).ready(function(){initNavTree('class_haplotype_scorer.html','');});
</script>
<div id="doc-content">
<!-- window showing the filter options -->
<div id="MSearchSelectWindow"
     onmouseover="return searchBox.OnSearchSelectShow()"
     onmouseout="return searchBox.OnSearchSelectHide()"
     onkeydown="return searchBox.OnSearchSelectKey(event)">
<a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(0)"><span class="SelectionMark">&#160;</span>All</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(1)"><span class="SelectionMark">&#160;</span>Classes</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(2)"><span class="SelectionMark">&#160;</span>Functions</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(3)"><span class="SelectionMark">&#160;</span>Variables</a><a class="SelectItem" href="javascript:void(0)" onclick="searchBox.OnSelectItem(4)"><span class="SelectionMark">&#160;</span>Pages</a></div>

<!-- iframe showing the search results (closed by default) -->
<div id="MSearchResultsWindow">
<iframe src="javascript:void(0)" frameborder="0" 
        name="MSearchResults" id="MSearchResults">
</iframe>
</div>

<div class="header">
  <div class="summary">
<a href="#pub-methods">Public Member Functions</a> &#124;
<a href="#pub-attribs">Public Attributes</a> &#124;
<a href="#pri-methods">Private Member Functions</a> &#124;
<a href="#pri-attribs">Private Attributes</a> &#124;
<a href="class_haplotype_scorer-members.html">List of all members</a>  </div>
  <div class="headertitle">
<div class="title">HaplotypeScorer Class Reference</div>  </div>
</div><!--header-->
<div class="contents">
<div class="dynheader">
Collaboration diagram for HaplotypeScorer:</div>
<div class="dyncontent">
<div class="center"><img src="class_haplotype_scorer__coll__graph.png" border="0" usemap="#_haplotype_scorer_coll__map" alt="Collaboration graph"/></div>
<map name="_haplotype_scorer_coll__map" id="_haplotype_scorer_coll__map">
<area shape="rect" id="node3" title="STL class. " alt="" coords="785,232,908,259"/><area shape="rect" id="node7" title="STL class. " alt="" coords="393,139,492,165"/><area shape="rect" id="node17" title="STL class. " alt="" coords="374,563,511,605"/><area shape="rect" id="node14" title="STL class. " alt="" coords="385,503,501,529"/><area shape="rect" id="node8" href="class_paired_read_mapper.html" title="PairedReadMapper" alt="" coords="1516,573,1649,600"/><area shape="rect" id="node9" href="class_sequence_graph.html" title="SequenceGraph" alt="" coords="1173,817,1288,844"/><area shape="rect" id="node11" href="class_node.html" title="Node" alt="" coords="416,872,469,899"/><area shape="rect" id="node12" title="STL class. " alt="" coords="184,723,264,749"/><area shape="rect" id="node13" title="STL class. " alt="" coords="6,715,133,757"/></map>
<center><span class="legend">[<a target="top" href="graph_legend.html">legend</a>]</span></center></div>
<table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-methods"></a>
Public Member Functions</h2></td></tr>
<tr class="memitem:a0a00aca26757cb7b0283d5306c96274c"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a0a00aca26757cb7b0283d5306c96274c"></a>
&#160;</td><td class="memItemRight" valign="bottom"><b>HaplotypeScorer</b> (<a class="el" href="class_sequence_graph.html">SequenceGraph</a> &amp;)</td></tr>
<tr class="separator:a0a00aca26757cb7b0283d5306c96274c"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a576c8ac9479fa28cd6703349e230e372"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a576c8ac9479fa28cd6703349e230e372"></a>
void&#160;</td><td class="memItemRight" valign="bottom"><b>find_possible_haplotypes</b> (int, std::string)</td></tr>
<tr class="separator:a576c8ac9479fa28cd6703349e230e372"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a2674aeb1990f180be417e4b819777b8e"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_haplotype_scorer.html#a2674aeb1990f180be417e4b819777b8e">load_haplotypes</a> (std::string, int)</td></tr>
<tr class="memdesc:a2674aeb1990f180be417e4b819777b8e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Load haplotypes, at this point from text file with hap contig per line, haps separated by 'next' - generate this with old phaser code.  <a href="#a2674aeb1990f180be417e4b819777b8e">More...</a><br/></td></tr>
<tr class="separator:a2674aeb1990f180be417e4b819777b8e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f785c2951cf84a37cf90953115497f8"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_haplotype_scorer.html#a8f785c2951cf84a37cf90953115497f8">count_barcode_votes</a> (std::string, std::string)</td></tr>
<tr class="memdesc:a8f785c2951cf84a37cf90953115497f8"><td class="mdescLeft">&#160;</td><td class="mdescRight">Map reads to graph, sum up matches for each comntig mapped to by each barcode barcode.  <a href="#a8f785c2951cf84a37cf90953115497f8">More...</a><br/></td></tr>
<tr class="separator:a8f785c2951cf84a37cf90953115497f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a677180428f96518e478f093ee2009a1e"><td class="memItemLeft" align="right" valign="top">int&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_haplotype_scorer.html#a677180428f96518e478f093ee2009a1e">score_haplotypes</a> ()</td></tr>
<tr class="memdesc:a677180428f96518e478f093ee2009a1e"><td class="mdescLeft">&#160;</td><td class="mdescRight">Decide winning phasing.  <a href="#a677180428f96518e478f093ee2009a1e">More...</a><br/></td></tr>
<tr class="separator:a677180428f96518e478f093ee2009a1e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaf914a1ee082d2a56eb15ee595862cf8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaf914a1ee082d2a56eb15ee595862cf8"></a>
int&#160;</td><td class="memItemRight" valign="bottom"><b>score_haplotypes2</b> ()</td></tr>
<tr class="separator:aaf914a1ee082d2a56eb15ee595862cf8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aa0fefe323ce39d294ec66593ef76eb9d"><td class="memItemLeft" align="right" valign="top">void&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_haplotype_scorer.html#aa0fefe323ce39d294ec66593ef76eb9d">decide_barcode_haplotype_support</a> ()</td></tr>
<tr class="memdesc:aa0fefe323ce39d294ec66593ef76eb9d"><td class="mdescLeft">&#160;</td><td class="mdescRight">sum each barcode's score for each haplotype  <a href="#aa0fefe323ce39d294ec66593ef76eb9d">More...</a><br/></td></tr>
<tr class="separator:aa0fefe323ce39d294ec66593ef76eb9d"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pub-attribs"></a>
Public Attributes</h2></td></tr>
<tr class="memitem:a5dd545475a60a0eca10720ee24c13929"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a5dd545475a60a0eca10720ee24c13929"></a>
std::map&lt; std::string, <br class="typebreak"/>
std::map&lt; sgNodeID_t, int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>barcode_node_mappings</b></td></tr>
<tr class="separator:a5dd545475a60a0eca10720ee24c13929"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a8f355d664516cea6a94a26c44a120748"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a8f355d664516cea6a94a26c44a120748"></a>
std::map&lt; std::string, <br class="typebreak"/>
std::map&lt; int, int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>barcode_haplotype_mappings</b></td></tr>
<tr class="separator:a8f355d664516cea6a94a26c44a120748"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:abd57cfc559f184d0de23a65382ca2732"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="abd57cfc559f184d0de23a65382ca2732"></a>
std::map&lt; std::string, <br class="typebreak"/>
std::map&lt; int, int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>barcode_haplotype_mappings2</b></td></tr>
<tr class="separator:abd57cfc559f184d0de23a65382ca2732"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-methods"></a>
Private Member Functions</h2></td></tr>
<tr class="memitem:ac303a8c2f375a3e11c35461f985ad2b1"><td class="memItemLeft" align="right" valign="top">std::vector&lt; int &gt;&#160;</td><td class="memItemRight" valign="bottom"><a class="el" href="class_haplotype_scorer.html#ac303a8c2f375a3e11c35461f985ad2b1">winner_for_barcode</a> (std::string barcode)</td></tr>
<tr class="memdesc:ac303a8c2f375a3e11c35461f985ad2b1"><td class="mdescLeft">&#160;</td><td class="mdescRight">Find haplotype that a barcode votes for.  <a href="#ac303a8c2f375a3e11c35461f985ad2b1">More...</a><br/></td></tr>
<tr class="separator:ac303a8c2f375a3e11c35461f985ad2b1"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table><table class="memberdecls">
<tr class="heading"><td colspan="2"><h2 class="groupheader"><a name="pri-attribs"></a>
Private Attributes</h2></td></tr>
<tr class="memitem:a59696d33cc73c33f06235dd83bb565d2"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a59696d33cc73c33f06235dd83bb565d2"></a>
<a class="el" href="class_sequence_graph.html">SequenceGraph</a> &amp;&#160;</td><td class="memItemRight" valign="bottom"><b>sg</b></td></tr>
<tr class="separator:a59696d33cc73c33f06235dd83bb565d2"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae01925c34fec59a564b796f6993222f8"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae01925c34fec59a564b796f6993222f8"></a>
<a class="el" href="class_paired_read_mapper.html">PairedReadMapper</a>&#160;</td><td class="memItemRight" valign="bottom"><b>mapper</b></td></tr>
<tr class="separator:ae01925c34fec59a564b796f6993222f8"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a153b4a4eeb24d76693d7ad242748a0ae"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a153b4a4eeb24d76693d7ad242748a0ae"></a>
std::set&lt; sgNodeID_t &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>haplotype_nodes</b></td></tr>
<tr class="separator:a153b4a4eeb24d76693d7ad242748a0ae"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ae4bc3c1fa1d830cd7740196f27660243"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ae4bc3c1fa1d830cd7740196f27660243"></a>
std::vector&lt; std::vector<br class="typebreak"/>
&lt; sgNodeID_t &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>haplotype_ids</b></td></tr>
<tr class="separator:ae4bc3c1fa1d830cd7740196f27660243"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:af6e66791251c186db4eeb6da358889fa"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="af6e66791251c186db4eeb6da358889fa"></a>
std::vector&lt; std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>unused_barcodes</b></td></tr>
<tr class="separator:af6e66791251c186db4eeb6da358889fa"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a91efa35e883030be8b9dbc77bbecfd35"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a91efa35e883030be8b9dbc77bbecfd35"></a>
std::map&lt; sgNodeID_t, <br class="typebreak"/>
std::vector&lt; int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>node_id_haplotype_index_map</b></td></tr>
<tr class="separator:a91efa35e883030be8b9dbc77bbecfd35"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:aaee5ac099a771364c9d8a4ffc881377e"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="aaee5ac099a771364c9d8a4ffc881377e"></a>
std::map&lt; sgNodeID_t, std::string &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>id_to_contig_name</b></td></tr>
<tr class="separator:aaee5ac099a771364c9d8a4ffc881377e"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:a77159aee7dacfa6900d39d97a461a94b"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="a77159aee7dacfa6900d39d97a461a94b"></a>
std::map&lt; int, std::map<br class="typebreak"/>
&lt; std::string, int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>haplotype_barcode_agree</b></td></tr>
<tr class="separator:a77159aee7dacfa6900d39d97a461a94b"><td class="memSeparator" colspan="2">&#160;</td></tr>
<tr class="memitem:ad252b59f691751759bd4b72a421b8252"><td class="memItemLeft" align="right" valign="top"><a class="anchor" id="ad252b59f691751759bd4b72a421b8252"></a>
std::map&lt; int, std::map<br class="typebreak"/>
&lt; std::string, int &gt; &gt;&#160;</td><td class="memItemRight" valign="bottom"><b>haplotype_barcode_disagree</b></td></tr>
<tr class="separator:ad252b59f691751759bd4b72a421b8252"><td class="memSeparator" colspan="2">&#160;</td></tr>
</table>
<a name="details" id="details"></a><h2 class="groupheader">Detailed Description</h2>
<div class="textblock">
<p>Definition at line <a class="el" href="_haplotype_scorer_8hpp_source.html#l00018">18</a> of file <a class="el" href="_haplotype_scorer_8hpp_source.html">HaplotypeScorer.hpp</a>.</p>
</div><h2 class="groupheader">Member Function Documentation</h2>
<a class="anchor" id="a8f785c2951cf84a37cf90953115497f8"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void HaplotypeScorer::count_barcode_votes </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>r1_filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>r2_filename</em>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Map reads to graph, sum up matches for each comntig mapped to by each barcode barcode. </p>
<dl class="section return"><dt>Returns</dt><dd>Load read files 1 and 2, populating map of id to barcode (this should be a vector the same as read to node but either my input reads are dodgy or there's something else wrong) find mappings for each node in the bubble contigs. for each barcode, sum up the number of mappings for each node that a read with that barcode maps to </dd></dl>

<p>Definition at line <a class="el" href="_haplotype_scorer_8cpp_source.html#l00130">130</a> of file <a class="el" href="_haplotype_scorer_8cpp_source.html">HaplotypeScorer.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="aa0fefe323ce39d294ec66593ef76eb9d"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void HaplotypeScorer::decide_barcode_haplotype_support </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>sum each barcode's score for each haplotype </p>
<dl class="section return"><dt>Returns</dt><dd>Returns a vector containing: total records generated by the factory, total records after filtering, total file records present in the file and the total number of records after filtering. </dd></dl>

<p>Definition at line <a class="el" href="_haplotype_scorer_8cpp_source.html#l00022">22</a> of file <a class="el" href="_haplotype_scorer_8cpp_source.html">HaplotypeScorer.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a2674aeb1990f180be417e4b819777b8e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">void HaplotypeScorer::load_haplotypes </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>haplotype_filename</em>, </td>
        </tr>
        <tr>
          <td class="paramkey"></td>
          <td></td>
          <td class="paramtype">int&#160;</td>
          <td class="paramname"><em>degree</em> = <code>2</code>&#160;</td>
        </tr>
        <tr>
          <td></td>
          <td>)</td>
          <td></td><td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Load haplotypes, at this point from text file with hap contig per line, haps separated by 'next' - generate this with old phaser code. </p>
<ul>
<li><dl class="section return"><dt>Returns</dt><dd>populate array eith possible haplotpyes: this should be every comination of bubble contigs </dd></dl>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000008">Todo:</a></b></dt><dd>add functionality to find bubbles on graph, and generate possible haplotypes </dd></dl>
</li>
</ul>

<p>Definition at line <a class="el" href="_haplotype_scorer_8cpp_source.html#l00082">82</a> of file <a class="el" href="_haplotype_scorer_8cpp_source.html">HaplotypeScorer.cpp</a>.</p>

</div>
</div>
<a class="anchor" id="a677180428f96518e478f093ee2009a1e"></a>
<div class="memitem">
<div class="memproto">
      <table class="memname">
        <tr>
          <td class="memname">int HaplotypeScorer::score_haplotypes </td>
          <td>(</td>
          <td class="paramname"></td><td>)</td>
          <td></td>
        </tr>
      </table>
</div><div class="memdoc">

<p>Decide winning phasing. </p>
<dl class="section return"><dt>Returns</dt><dd>Calculate, for each haplotype, a number of metrics that represent how well the 10x reads support that phasing loop over each barcode in turn, then each phasing pair </dd></dl>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000009">Todo:</a></b></dt><dd>decide criteria for winning haploype</dd></dl>

<p>Definition at line <a class="el" href="_haplotype_scorer_8cpp_source.html#l00157">157</a> of file <a class="el" href="_haplotype_scorer_8cpp_source.html">HaplotypeScorer.cpp</a>.</p>

<p><div class="dynheader">
Here is the call graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_haplotype_scorer_a677180428f96518e478f093ee2009a1e_cgraph.png" border="0" usemap="#class_haplotype_scorer_a677180428f96518e478f093ee2009a1e_cgraph" alt=""/></div>
<map name="class_haplotype_scorer_a677180428f96518e478f093ee2009a1e_cgraph" id="class_haplotype_scorer_a677180428f96518e478f093ee2009a1e_cgraph">
<area shape="rect" id="node2" href="class_haplotype_scorer.html#ac303a8c2f375a3e11c35461f985ad2b1" title="Find haplotype that a barcode votes for. " alt="" coords="213,5,379,46"/></map>
</div>
</p>

</div>
</div>
<a class="anchor" id="ac303a8c2f375a3e11c35461f985ad2b1"></a>
<div class="memitem">
<div class="memproto">
<table class="mlabels">
  <tr>
  <td class="mlabels-left">
      <table class="memname">
        <tr>
          <td class="memname">std::vector&lt; int &gt; HaplotypeScorer::winner_for_barcode </td>
          <td>(</td>
          <td class="paramtype">std::string&#160;</td>
          <td class="paramname"><em>barcode</em></td><td>)</td>
          <td></td>
        </tr>
      </table>
  </td>
  <td class="mlabels-right">
<span class="mlabels"><span class="mlabel">private</span></span>  </td>
  </tr>
</table>
</div><div class="memdoc">

<p>Find haplotype that a barcode votes for. </p>
<ul>
<li><dl class="section return"><dt>Returns</dt><dd>loop over barcode aplotype mappings to find maximum support if more than2 have same support both are returned </dd></dl>
<dl class="todo"><dt><b><a class="el" href="todo.html#_todo000010">Todo:</a></b></dt><dd>decide heuristics to vote for winner </dd></dl>
</li>
</ul>

<p>Definition at line <a class="el" href="_haplotype_scorer_8cpp_source.html#l00242">242</a> of file <a class="el" href="_haplotype_scorer_8cpp_source.html">HaplotypeScorer.cpp</a>.</p>

<p><div class="dynheader">
Here is the caller graph for this function:</div>
<div class="dyncontent">
<div class="center"><img src="class_haplotype_scorer_ac303a8c2f375a3e11c35461f985ad2b1_icgraph.png" border="0" usemap="#class_haplotype_scorer_ac303a8c2f375a3e11c35461f985ad2b1_icgraph" alt=""/></div>
<map name="class_haplotype_scorer_ac303a8c2f375a3e11c35461f985ad2b1_icgraph" id="class_haplotype_scorer_ac303a8c2f375a3e11c35461f985ad2b1_icgraph">
<area shape="rect" id="node2" href="class_haplotype_scorer.html#a677180428f96518e478f093ee2009a1e" title="Decide winning phasing. " alt="" coords="219,5,378,46"/></map>
</div>
</p>

</div>
</div>
<hr/>The documentation for this class was generated from the following files:<ul>
<li>/home/travis/build/bioinfologics/bsg/src/sglib/<a class="el" href="_haplotype_scorer_8hpp_source.html">HaplotypeScorer.hpp</a></li>
<li>/home/travis/build/bioinfologics/bsg/src/sglib/<a class="el" href="_haplotype_scorer_8cpp_source.html">HaplotypeScorer.cpp</a></li>
</ul>
</div><!-- contents -->
</div><!-- doc-content -->
<!-- start footer part -->
<div id="nav-path" class="navpath"><!-- id is needed for treeview function! -->
  <ul>
    <li class="navelem"><a class="el" href="class_haplotype_scorer.html">HaplotypeScorer</a></li>
    <li class="footer">Generated by
    <a href="http://www.doxygen.org/index.html">
    <img class="footer" src="doxygen.png" alt="doxygen"/></a> 1.8.6 </li>
  </ul>
</div>
</body>
</html>
